src/MetadataManager/PathToHash.ts

```diff
@@ -0,0 +1,269 @@
+   */
+  get(uri: vscode.Uri) {
+    const splitPath = vscode.workspace.asRelativePath(uri).split('/');
+    let map = this._map;
+
+    splitPath.forEach((path) => {
+      if (map === undefined) {
+        return undefined;
+      }
+      map = map[path];
+    });
+
+    return map;
+  }
+
+  getFilesUnderFolder(uri: vscode.Uri) {
+    const folder = this.get(uri);
+    const files: vscode.Uri[] = [];
+    if (typeof (folder) === 'string') {
+      // not a folder
+      return files;
+    }
+    for (const name in folder) {
+      files.push(vscode.Uri.joinPath(uri, name));
+    }
+
+    return files;
+  }
+
+  /**
+   * @brief Input uri, extract path and hash values, and store them in a '_map'
+   */  
+  async add(uri: vscode.Uri) {
+    const path = vscode.workspace.asRelativePath(uri);
+    const paths = path.split('/');
+    let content: any = await PathToHash.generateHash(uri);
+    let subMap = this._map;
+    let idx = 0;
+    for (let path = paths[idx]; idx < paths.length - 1; path = paths[++idx]) {
+      if (!subMap[path]) {
+        break;
+      }
+      subMap = subMap[path];
+    }
+    if (paths.length - 1 === idx) {  // paths.length - 1: index of a file name
+      // When all of the folder path are stored in pathToHash
+      // update / create pathToHash for a file
+      subMap[paths[idx]] = content;
+      return;
+    }
+
+    for (let i = paths.length - 1; i > idx; --i) {
+      let tempContent: {[key: string]: any} = {};
+      tempContent[paths[i]] = content;
+      content = tempContent;
+    }
+    subMap[paths[idx]] = content;
+}
+
+  /**
+   * @param uri 
+   * @brief Input uri, extract path and hash values, and delete them in a '_map'
+   */
+  delete(uri: vscode.Uri) {
+    let subMap = this._map;
+    const splitPath = vscode.workspace.asRelativePath(uri).split('/');
+
+    for (let i = 0, path = splitPath[i]; i < splitPath.length - 1; path = splitPath[++i]) {
+      if (!subMap) {
+        return;
+      }
+      subMap = subMap[path];
+    }
+    if (subMap === undefined) {
+      // already deleted
+      return;
+    }
+    delete subMap[splitPath[splitPath.length - 1]];
+    if (splitPath.length > 1) {
+      this.deleteEmptyDirPath(this._map, splitPath, 0);
+    }
+  }
+
+  /**
+   * 
+   * @param parent 
+   * @param paths 
+   * @param idx 
+   * @brief delete Empty directory path from '_map'
+   */
+  deleteEmptyDirPath(parent: any, paths: string[], idx: number) {
+    const path = paths[idx];
+    if (paths.length - 2 === idx) {
+      if (Object.keys(parent[path]).length === 0) {
+        delete parent[path];
+      }
+      return;
+    }
+    if (parent[path] === undefined) {
+      return;
+    }
+    this.deleteEmptyDirPath(parent[path], paths, idx + 1);
+    if (Object.keys(parent[path]).length === 0) {
+      delete parent[path];
+    }
+  }
+
+  // this function will probably be separated by Util
+  public static async generateHash(uri: vscode.Uri) {
```

Comment:
Could you extract outside the class, (moving this into `Utility` directory will be the best, but if you're keeing this function inside the file)?

Reply:
I deleted this

Reply:
@gandi0330 Isn't this the newest version..?